src/NSubstitute/Core/Extensions.cs

```diff
@@ -24,17 +24,44 @@ public static bool IsCompatibleWith(this object? instance, Type type)
         }
 
         var instanceType = instance.GetType();
+        var compatibleInstanceTypes = GetCompatibleTypes(instanceType);
 
-        if (instanceType.IsGenericType && type.IsGenericType
-                && instanceType.GetGenericTypeDefinition() == type.GetGenericTypeDefinition())
+        foreach (var aCompatibleInstanceType in compatibleInstanceTypes)
         {
-            // both are the same generic type. If their GenericTypeArguments match then they are equivalent 
-            return CallSpecification.TypesAreAllEquivalent(instanceType.GenericTypeArguments, type.GenericTypeArguments);
+            if (aCompatibleInstanceType.IsGenericType &&
+                type.IsGenericType &&
+                aCompatibleInstanceType.GetGenericTypeDefinition() == type.GetGenericTypeDefinition())
+            {
+                // both are the same generic type. If their GenericTypeArguments match then they are equivalent
+                return CallSpecification.TypesAreAllEquivalent(
+                    aCompatibleInstanceType.GenericTypeArguments, type.GenericTypeArguments);
+            }
         }
 
         return requiredType.IsInstanceOfType(instance);
     }
 
+    private static IReadOnlyList<Type> GetCompatibleTypes(Type type)
```

Comment:
**question:** I'm not sure if the benchmark project still works (it hasn't been maintained for a while), but if so it would be great to get a perf measurement of this change. It seems potentially expensive?

Is this mainly for `Arg.AnyType` support? Maybe we can optimise for that case?

(Please let me know if i'm being too paranoid about the impact here 😅 )

Reply:
Indeed, the intuition that only **Arg.AnyType**-cases need additional support has proven to be correct.

I have added several tests to capture various inheritance and argument matching scenarios, and the only tests not succeeding with the old implementation have been the ones with the generic interface's generic type argument bound to **Arg.AnyType**.

Thus there was no need for a complicated recursive **GetCompatibleTypes()** method with potential performance issues, a single level of depth has shown itself to be sufficient.

Reply:
Changes look great 👍 Much simpler 😄